crates/gen-guest-teavm-java/src/lib.rs

```diff
@@ -0,0 +1,1918 @@
+
+            Instruction::Malloc { .. } => unimplemented!(),
+
+            Instruction::GuestDeallocate { size, align } => {
+                uwriteln!(
+                    self.src,
+                    "Memory.free(Address.fromInt({}), {size}, {align});",
+                    operands[0]
+                )
+            }
+
+            Instruction::GuestDeallocateString => uwriteln!(
+                self.src,
+                "Memory.free(Address.fromInt({}), {}, 1);",
+                operands[0],
+                operands[1]
+            ),
+
+            Instruction::GuestDeallocateVariant { blocks } => {
+                let cases = self
+                    .blocks
+                    .drain(self.blocks.len() - blocks..)
+                    .enumerate()
+                    .map(|(i, Block { body, results, .. })| {
+                        assert!(results.is_empty());
+
+                        format!(
+                            "case {i}: {{
+                                 {body}
+                                 break;
+                             }}"
+                        )
+                    })
+                    .collect::<Vec<_>>()
+                    .join("\n");
+
+                let op = &operands[0];
+
+                uwrite!(
+                    self.src,
+                    "
+                    switch ({op}) {{
+                        {cases}
+                    }}
+                    "
+                );
+            }
+
+            Instruction::GuestDeallocateList { element } => {
+                let Block {
+                    body,
+                    results,
+                    base,
+                    ..
+                } = self.blocks.pop().unwrap();
+                assert!(results.is_empty());
+
+                let address = &operands[0];
+                let length = &operands[1];
+
+                let size = self.gen.sizes.size(element);
+                let align = self.gen.sizes.align(element);
+
+                if !body.trim().is_empty() {
+                    let index = self.locals.tmp("index");
+
+                    uwrite!(
+                        self.src,
+                        "
+                        for (int {index} = 0; {index} < ({length}); ++{index}) {{
+                            int {base} = ({address}) + ({index} * {size});
+                            {body}
+                        }}
+                        "
+                    );
+                }
+
+                uwriteln!(
+                    self.src,
+                    "Memory.free(Address.fromInt({address}), ({length}) * {size}, {align});"
+                );
+            }
+        }
+    }
+
+    fn return_pointer(&mut self, _iface: &Interface, size: usize, align: usize) -> String {
+        self.gen.return_area_size = self.gen.return_area_size.max(size);
+        self.gen.return_area_align = self.gen.return_area_align.max(align);
+        "RETURN_AREA".into()
+    }
```

Comment:
One change we just landed to the C & Rust generators was that for imports the return area was stored on the stack, but I'm not sure if the equivalent would work in java. That being said it may be worth considering if you'd like.

Reply:
Yeah, I saw that.  There's no shadow stack in TeaVM currently (and I don't know if it would make sense given you can't take the address of a stack-allocated thing in Java), so literally putting it on the stack isn't an option, but we could certainly allocate it on-the-fly on the GC'd heap.

Reply:
Correction: TeaVM _does_ have a shadow stack, since the GC needs to scan it for roots when collecting.  I don't know enough about how it works yet to safely allocate a return area on it, and I suspect TeaVM expects only heap references on that stack, so I think it would be a bad fit for this kind of thing.

Also, forget my idea of allocating the return area on the GC'd heap since TeaVM may consider it garbage once there are no more stack references to it.

Anyway, the static allocation works for now, and we can revisit if/when we have to deal with reentrancy.

Reply:
One more thought: I just realized that the C and Rust generators only use the stack for imports (which is what you said, now that I read more carefully).  So yes, we could at least use an on-the-fly heap allocation for import return areas, either using the GC'd heap or the manually managed one.  The static allocation is best for performance as long as reentrancy isn't a concern, though.  And as with C and Rust, static allocation is the only option for exports.